File: backend/app/services/export_service.py
from sqlalchemy.orm import Session
from app.models import (
    Role, Company, Accomplishment, AccomplishmentCategory,
    Award, Presentation, Profile, Education, Certification,
    Skill, RoleSkill, TeamStructure, Responsibility,
)


def generate_export_document(
    db: Session,
    role_ids: list[int],
    include_supporting: bool = True,
    include_awards: bool = True,
    include_presentations: bool = True,
    include_responsibilities: bool = True,
) -> str:
    sections = []

    # Profile
    profile = db.query(Profile).first()
    if profile:
        sections.append("# Profile")
        sections.append(f"**Name:** {profile.name}")
        if profile.email:
            sections.append(f"**Email:** {profile.email}")
        if profile.phone:
            sections.append(f"**Phone:** {profile.phone}")
        if profile.city and profile.state:
            sections.append(f"**Location:** {profile.city}, {profile.state}")
        if profile.portfolio_url:
            sections.append(f"**Portfolio:** {profile.portfolio_url}")
        if profile.linkedin_url:
            sections.append(f"**LinkedIn:** {profile.linkedin_url}")
        sections.append("")

    # Roles grouped by company
    roles = db.query(Role).filter(Role.id.in_(role_ids)).order_by(Role.start_date.desc()).all()
    company_roles: dict[int, list[Role]] = {}
    for role in roles:
        company_roles.setdefault(role.company_id, []).append(role)

    sections.append("# Work Experience")
    sections.append("")

    for company_id, company_role_list in company_roles.items():
        company = db.query(Company).filter(Company.id == company_id).first()
        if not company:
            continue

        date_range = _format_date_range(company.start_date, company.end_date)
        sections.append(f"## {company.name} ({date_range})")
        if company.description:
            sections.append(f"*{company.description}*")
        if company.industry:
            sections.append(f"Industry: {company.industry}")
        sections.append("")

        for role in company_role_list:
            role_date_range = _format_date_range(role.start_date, role.end_date)
            location = _format_location(role.city, role.state, role.is_remote)
            sections.append(f"### {role.title} ({role_date_range}) — {location}")
            if role.summary:
                sections.append(f"\n{role.summary}")
            sections.append("")

            # Team structure context
            if role.team_structure:
                ts = role.team_structure
                team_info = []
                if ts.direct_reports:
                    team_info.append(f"{ts.direct_reports} direct reports")
                if ts.team_size:
                    team_info.append(f"team of {ts.team_size}")
                if ts.responsibilities:
                    team_info.append(ts.responsibilities)
                if team_info:
                    sections.append(f"**Team Context:** {'; '.join(team_info)}")
                    sections.append("")

            # Responsibilities
            if include_responsibilities:
                responsibilities = (
                    db.query(Responsibility)
                    .filter(Responsibility.role_id == role.id)
                    .order_by(Responsibility.sort_order)
                    .all()
                )
                if responsibilities:
                    sections.append("**Key Responsibilities:**")
                    for resp in responsibilities:
                        sections.append(f"- {resp.description}")
                    sections.append("")

            # Accomplishments - resume bullets
            resume_bullets = (
                db.query(Accomplishment)
                .filter(
                    Accomplishment.role_id == role.id,
                    Accomplishment.category == AccomplishmentCategory.RESUME_BULLET,
                )
                .order_by(Accomplishment.sort_order)
                .all()
            )
            if resume_bullets:
                sections.append("**Key Accomplishments:**")
                for acc in resume_bullets:
                    sections.append(f"- {acc.description}")
                    if acc.tech_stack:
                        sections.append(f"  - Tech Stack: {acc.tech_stack}")
                    if acc.key_result:
                        sections.append(f"  - Key Result: {acc.key_result}")
                sections.append("")

            # Supporting details
            if include_supporting:
                supporting = (
                    db.query(Accomplishment)
                    .filter(
                        Accomplishment.role_id == role.id,
                        Accomplishment.category == AccomplishmentCategory.SUPPORTING_DETAIL,
                    )
                    .order_by(Accomplishment.sort_order)
                    .all()
                )
                if supporting:
                    sections.append("**Supporting Details (additional context for AI):**")
                    for acc in supporting:
                        sections.append(f"- {acc.description}")
                        if acc.tech_stack:
                            sections.append(f"  - Tech Stack: {acc.tech_stack}")
                        if acc.key_result:
                            sections.append(f"  - Key Result: {acc.key_result}")
                    sections.append("")

            # Skills for this role
            role_skills = db.query(RoleSkill).filter(RoleSkill.role_id == role.id).all()
            if role_skills:
                skill_names = []
                for rs in role_skills:
                    skill = db.query(Skill).filter(Skill.id == rs.skill_id).first()
                    if skill:
                        entry = skill.name
                        if rs.proficiency:
                            entry += f" ({rs.proficiency})"
                        skill_names.append(entry)
                sections.append(f"**Skills:** {', '.join(skill_names)}")
                sections.append("")

            # Awards
            if include_awards:
                awards = (
                    db.query(Award)
                    .filter(Award.role_id == role.id, Award.resume_relevant.is_(True))
                    .all()
                )
                if awards:
                    sections.append("**Awards:**")
                    for award in awards:
                        line = f"- {award.title}"
                        if award.issuer:
                            line += f" ({award.issuer})"
                        if award.date:
                            line += f" — {award.date.strftime('%b %Y')}"
                        sections.append(line)
                        if award.description:
                            sections.append(f"  {award.description}")
                    sections.append("")

            # Presentations
            if include_presentations:
                presentations = (
                    db.query(Presentation)
                    .filter(Presentation.role_id == role.id, Presentation.resume_relevant.is_(True))
                    .all()
                )
                if presentations:
                    sections.append("**Presentations:**")
                    for pres in presentations:
                        line = f"- {pres.title}"
                        if pres.venue:
                            line += f" at {pres.venue}"
                        if pres.date:
                            line += f" ({pres.date.strftime('%b %Y')})"
                        if pres.audience:
                            line += f" — Audience: {pres.audience}"
                        sections.append(line)
                    sections.append("")

    # Education
    education_list = db.query(Education).all()
    if education_list:
        sections.append("# Education")
        for edu in education_list:
            line = f"**{edu.institution}**"
            if edu.degree and edu.field:
                line += f" — {edu.degree} in {edu.field}"
            elif edu.degree:
                line += f" — {edu.degree}"
            if edu.city and edu.state:
                line += f" ({edu.city}, {edu.state})"
            sections.append(line)
        sections.append("")

    # Certifications
    certs = db.query(Certification).all()
    if certs:
        sections.append("# Certifications")
        for cert in certs:
            line = f"- **{cert.name}**"
            if cert.issuing_org:
                line += f" — {cert.issuing_org}"
            if cert.issue_date:
                line += f" ({cert.issue_date.strftime('%b %Y')})"
            sections.append(line)
        sections.append("")

    return "\n".join(sections)


def _format_date_range(start, end) -> str:
    parts = []
    if start:
        parts.append(start.strftime("%b %Y"))
    if end:
        parts.append(end.strftime("%b %Y"))
    else:
        parts.append("Present")
    return " - ".join(parts) if parts else "N/A"


def _format_location(city, state, is_remote) -> str:
    location = ""
    if city and state:
        location = f"{city}, {state}"
    if is_remote:
        location = f"Remote ({location})" if location else "Remote"
    return location or "N/A"

File: server/src/services/exportService.ts
import { eq, inArray, asc } from 'drizzle-orm';
import { db } from '../db/index.js';
import {
  profiles, companies, roles, accomplishments, responsibilities,
  role_skills, skills, awards, presentations, team_structures,
  education, certifications,
} from '../db/schema.js';

function formatDateRange(start: string | null, end: string | null): string {
  const parts: string[] = [];
  if (start) parts.push(formatMonthYear(start));
  if (end) parts.push(formatMonthYear(end));
  else parts.push('Present');
  return parts.length > 0 ? parts.join(' - ') : 'N/A';
}

function formatMonthYear(dateStr: string): string {
  const d = new Date(dateStr + 'T00:00:00');
  return d.toLocaleDateString('en-US', { month: 'short', year: 'numeric' });
}

function formatLocation(city: string | null, state: string | null, isRemote: boolean | null): string {
  let location = '';
  if (city && state) location = `${city}, ${state}`;
  if (isRemote) location = location ? `Remote (${location})` : 'Remote';
  return location || 'N/A';
}

export function generateExportDocument(
  roleIds: number[],
  includingSupporting = false,
  includeAwards = false,
  includePresentations = false,
  includeResponsibilities = true,
  includeTeamNarrative = false,
): string {
  const sections: string[] = [];

  // Profile
  const profile = db.select().from(profiles).limit(1).all()[0];
  if (profile) {
    sections.push('# Profile');
    sections.push(`**Name:** ${profile.name}`);
    if (profile.email) sections.push(`**Email:** ${profile.email}`);
    if (profile.phone) sections.push(`**Phone:** ${profile.phone}`);
    if (profile.city && profile.state) sections.push(`**Location:** ${profile.city}, ${profile.state}`);
    if (profile.portfolio_url) sections.push(`**Portfolio:** ${profile.portfolio_url}`);
    if (profile.linkedin_url) sections.push(`**LinkedIn:** ${profile.linkedin_url}`);
    sections.push('');
  }

  // Roles grouped by company
  const roleRows = roleIds.length > 0
    ? db.select().from(roles).where(inArray(roles.id, roleIds))
        .orderBy(asc(roles.start_date)).all()
        .sort((a, b) => {
          if (!a.start_date) return 1;
          if (!b.start_date) return -1;
          return a.start_date < b.start_date ? 1 : -1;
        })
    : [];

  const companyRoles: Record<number, typeof roleRows> = {};
  for (const role of roleRows) {
    if (!companyRoles[role.company_id]) companyRoles[role.company_id] = [];
    companyRoles[role.company_id].push(role);
  }

  sections.push('# Work Experience');
  sections.push('');

  for (const [companyIdStr, companyRoleList] of Object.entries(companyRoles)) {
    const companyId = parseInt(companyIdStr);
    const company = db.select().from(companies).where(eq(companies.id, companyId)).limit(1).all()[0];
    if (!company) continue;

    const dateRange = formatDateRange(company.start_date, company.end_date);
    sections.push(`## ${company.name} (${dateRange})`);
    if (company.description) sections.push(`*${company.description}*`);
    if (company.industry) sections.push(`Industry: ${company.industry}`);
    sections.push('');

    for (const role of companyRoleList) {
      const roleRange = formatDateRange(role.start_date, role.end_date);
      const location = formatLocation(role.city, role.state, role.is_remote);
      sections.push(`### ${role.title} (${roleRange}) — ${location}`);
      if (role.summary) sections.push(`\n${role.summary}`);
      sections.push('');

      // Team structure
      const ts = db.select().from(team_structures).where(eq(team_structures.role_id, role.id)).limit(1).all()[0];
      if (ts) {
        const compactParts: string[] = [];
        if (ts.direct_reports) compactParts.push(`${ts.direct_reports} direct reports`);
        if (ts.team_size) compactParts.push(`team of ${ts.team_size}`);
        if (compactParts.length > 0) {
          sections.push(`**Team Context:** ${compactParts.join('; ')}`);
          sections.push('');
        }
        if (includeTeamNarrative && ts.responsibilities) {
          sections.push('**Team Structure Detail:**');
          sections.push(ts.responsibilities);
          sections.push('');
        }
      }

      // Responsibilities
      if (includeResponsibilities) {
        const resps = db.select().from(responsibilities)
          .where(eq(responsibilities.role_id, role.id))
          .orderBy(asc(responsibilities.sort_order))
          .all();
        if (resps.length > 0) {
          sections.push('**Key Responsibilities:**');
          for (const resp of resps) sections.push(`- ${resp.description}`);
          sections.push('');
        }
      }

      // Resume bullets
      const bullets = db.select().from(accomplishments)
        .where(eq(accomplishments.role_id, role.id))
        .orderBy(asc(accomplishments.sort_order))
        .all()
        .filter(a => a.category?.toUpperCase() === 'RESUME_BULLET');
      if (bullets.length > 0) {
        sections.push('**Key Accomplishments:**');
        for (const acc of bullets) {
          sections.push(`- ${acc.description}`);
          if (acc.tech_stack) sections.push(`  - Tech Stack: ${acc.tech_stack}`);
          if (acc.key_result) sections.push(`  - Key Result: ${acc.key_result}`);
        }
        sections.push('');
      }

      // Supporting details
      if (includingSupporting) {
        const supporting = db.select().from(accomplishments)
          .where(eq(accomplishments.role_id, role.id))
          .orderBy(asc(accomplishments.sort_order))
          .all()
          .filter(a => a.category?.toUpperCase() === 'SUPPORTING_DETAIL');
        if (supporting.length > 0) {
          sections.push('**Supporting Details (additional context for AI):**');
          for (const acc of supporting) {
            sections.push(`- ${acc.description}`);
            if (acc.tech_stack) sections.push(`  - Tech Stack: ${acc.tech_stack}`);
            if (acc.key_result) sections.push(`  - Key Result: ${acc.key_result}`);
          }
          sections.push('');
        }
      }

      // Skills for this role
      const roleSkillRows = db.select().from(role_skills)
        .where(eq(role_skills.role_id, role.id))
        .all();
      if (roleSkillRows.length > 0) {
        const skillNames: string[] = [];
        for (const rs of roleSkillRows) {
          const skill = db.select().from(skills).where(eq(skills.id, rs.skill_id)).limit(1).all()[0];
          if (skill) {
            skillNames.push(rs.proficiency ? `${skill.name} (${rs.proficiency})` : skill.name);
          }
        }
        if (skillNames.length > 0) {
          sections.push(`**Skills:** ${skillNames.join(', ')}`);
          sections.push('');
        }
      }

      // Awards
      if (includeAwards) {
        const awardRows = db.select().from(awards)
          .where(eq(awards.role_id, role.id))
          .all()
          .filter(a => a.resume_relevant);
        if (awardRows.length > 0) {
          sections.push('**Awards:**');
          for (const award of awardRows) {
            let line = `- ${award.title}`;
            if (award.issuer) line += ` (${award.issuer})`;
            if (award.date) line += ` — ${formatMonthYear(award.date)}`;
            sections.push(line);
            if (award.description) sections.push(`  ${award.description}`);
          }
          sections.push('');
        }
      }

      // Presentations
      if (includePresentations) {
        const presRows = db.select().from(presentations)
          .where(eq(presentations.role_id, role.id))
          .all()
          .filter(p => p.resume_relevant);
        if (presRows.length > 0) {
          sections.push('**Presentations:**');
          for (const pres of presRows) {
            let line = `- ${pres.title}`;
            if (pres.venue) line += ` at ${pres.venue}`;
            if (pres.date) line += ` (${formatMonthYear(pres.date)})`;
            if (pres.audience) line += ` — Audience: ${pres.audience}`;
            sections.push(line);
          }
          sections.push('');
        }
      }
    }
  }

  // Education
  const eduRows = db.select().from(education).all();
  if (eduRows.length > 0) {
    sections.push('# Education');
    for (const edu of eduRows) {
      let line = `**${edu.institution}**`;
      if (edu.degree && edu.field) line += ` — ${edu.degree} in ${edu.field}`;
      else if (edu.degree) line += ` — ${edu.degree}`;
      if (edu.city && edu.state) line += ` (${edu.city}, ${edu.state})`;
      sections.push(line);
    }
    sections.push('');
  }

  // Certifications
  const certRows = db.select().from(certifications).all();
  if (certRows.length > 0) {
    sections.push('# Certifications');
    for (const cert of certRows) {
      let line = `- **${cert.name}**`;
      if (cert.issuing_org) line += ` — ${cert.issuing_org}`;
      if (cert.issue_date) line += ` (${formatMonthYear(cert.issue_date)})`;
      sections.push(line);
    }
    sections.push('');
  }

  return sections.join('\n');
}
